src/matrices.jl

```diff
@@ -219,23 +219,54 @@ end
 
 # If `scale` is a function, compute the numeric value of the scale based on the
 # distance matrix; otherwise return the value of `scale` itself
-_computescale(scale::Function, x, y, metric) = scale(distancematrix(x, y, metric))
 _computescale(scale::Real, args...) = scale
+function _computescale(scale::Function, x, y, metric)
+    if x===y
+        distances = zeros(Int(length(x)*(length(x)-1)/2))
+        c = 0
+        @inbounds for (i,xi) in enumerate(x), (j,yj) in enumerate(y)
+            if j > i
+                distances[c+=1] = evaluate(metric, xi, yj)
+            end
+        end
+    else
+        distances = distancematrix(x, y, metric)
+    end
+    return scale(distances)
+end
 # specific methods to avoid `distancematrix`
 function _computescale(scale::typeof(maximum), x, y, metric::Metric)
     maxvalue = zero(eltype(x))
-    @inbounds for xi in x, yj in y
-        newvalue = evaluate(metric, xi, yj)
-        (newvalue > maxvalue) && (maxvalue = newvalue)
+    if x===y
+        @inbounds for (i,xi) in enumerate(x), (j,yj) in enumerate(y)
```

Comment:
I wonder. Can't this loop be reduced even further, to not require 2 full loops? I think we can do something like:
```julia
for i in 1:length(x)-1  
  for j in (i+1):length(x)
    @inbounds newvalue = evaluate(metric, x[i], y[j])
     (newvalue > maxvalue) && (maxvalue = newvalue)
  end
end
```

I think the above avoids the boolean comparison alltogether. But Is it right? I just pulled out of the top of my head :P